scene_graph/vrd_tutorial.py

```diff
@@ -0,0 +1,335 @@
+# %%
+from snorkel.labeling.model import LabelModel
+
+label_model = LabelModel(cardinality=3, verbose=True)
+label_model.fit(L_train, seed=123, lr=0.01, log_freq=10, n_epochs=100)
+
+# %%
+label_model.score(L_valid, Y_valid, metrics=["f1_micro"])
+
+# %% [markdown]
+# ## 4. Train a Classifier
+# You can then use these training labels to train any standard discriminative model, such as [a state-of-the-art ResNet](https://github.com/KaimingHe/deep-residual-networks), which should learn to generalize beyond the LF's we've developed!
+
+# %% [markdown]
+# #### Create DataLoaders for Classifier
+
+# %%
+from snorkel.classification.data import DictDataLoader
+from scene_graph.model import FlatConcat, SceneGraphDataset, WordEmb, init_fc
+
+#change to "scene_graph/data/VRD/sg_dataset/sg_train_images" for full set
+TRAIN_DIR = "scene_graph/data/VRD/sg_dataset/samples" 
+train_df["labels"] = label_model.predict(L_train)
+
+train_dl = DictDataLoader(
+    SceneGraphDataset("train_dataset", "train", TRAIN_DIR, train_df),
+    batch_size=16,
+    shuffle=True,
+)
+
+valid_dl = DictDataLoader(
+    SceneGraphDataset("valid_dataset", "valid", TRAIN_DIR, valid_df),
+    batch_size=16,
+    shuffle=False,
+)
+
+# %% [markdown]
+# #### Define Loss Functions
+
+
+# %%
+import torch.nn.functional as F
+
+def ce_loss(module_name, outputs, Y, active):
+    return F.cross_entropy(outputs[module_name][0][active], (Y.view(-1))[active])
+
+
+def softmax(module_name, outputs):
+    return F.softmax(outputs[module_name][0], dim=1)
+
+
+# %% [markdown]
+# #### Define Model Architecture
+
+# %%
+import torchvision.models as models
+import torch.nn as nn
+
+from functools import partial
+from snorkel.classification.scorer import Scorer
+from snorkel.classification.task import Operation, Task
+
+# initialize pretrained feature extractor
+cnn = models.resnet18(pretrained=True)
+
+# freeze the resnet weights
+for param in cnn.parameters():
+    param.requires_grad = False
+
+# define input features
+in_features = cnn.fc.in_features
+feature_extractor = nn.Sequential(*list(cnn.children())[:-1])
+
+# initialize FC layer: maps 3 sets of image features to class logits
+WEMB_SIZE = 100
+fc = nn.Linear(in_features * 3 + 2 * WEMB_SIZE, 3)
+init_fc(fc)
+
+# define layers
+module_pool = nn.ModuleDict(
+    {
+        "feat_extractor": feature_extractor,
+        "prediction_head": fc,
+        "feat_concat": FlatConcat(),
+        "word_emb": WordEmb(),
+    }
+)
+
+union_feat_op = Operation(
+    name="union_feat_op",
+    module_name="feat_extractor",
+    inputs=[("_input_", "union_crop")],
+)
+
+sub_feat_op = Operation(
+    name="sub_feat_op", module_name="feat_extractor", inputs=[("_input_", "sub_crop")]
+)
+
+obj_feat_op = Operation(
+    name="obj_feat_op", module_name="feat_extractor", inputs=[("_input_", "obj_crop")]
+)
+
+word_emb_op = Operation(
+    name="word_emb_op",
+    module_name="word_emb",
+    inputs=[("_input_", "sub_category"), ("_input_", "obj_category")],
+)
+
```

Comment:
comment: # define an operation to concatenate image features and word embeddings